core/src/main/java/org/mule/runtime/core/DefaultMuleEvent.java

```diff
@@ -990,6 +905,35 @@ public void setSecurityContext(SecurityContext context) {
     session.setSecurityContext(context);
   }
 
+  private MuleEvent parent;
+  private Correlation correlation;
+
+  public void setParent(MuleEvent parent) {
+    this.parent = parent;
+  }
+
+  @Override
+  public MuleEvent getParent() {
+    return parent;
+  }
+
+  public void setCorrelation(Correlation correlation) {
+    this.correlation = correlation;
+  }
+
+  @Override
+  public Correlation getCorrelation() {
+    return correlation;
+  }
+
+  @Override
+  public String getCorrelationId() {
+    return getCorrelation().getId().orElse(getExecutionContext().getCorrelationId().orElse(getMessage().getUniqueId())
```

Comment:
For performance this should be fixed value defined when event is build.  But this can be done along with work on EventBuilder.